trunk/SUAVE/Components/Energy/Converters/Propeller.py

```diff
@@ -381,7 +348,102 @@ def spin(self,conditions):
         Va_ind_avg = va
         Vt_avg     = Wt
         Va_avg     = Wa
-          
+        
+        #------------------------------------------------------------------------------------------
+        # Now, if nonuniform freestream, analyze with the nonuniform velocities at propeller blade sections:
+        #------------------------------------------------------------------------------------------
+        
+        if nonuniform_freestream:             
+            
+            # velocities at each polar grid position on propeller disc with shape = (ctrl_pts,Na,Nr)
+            Vt_2d = self.tangential_velocities_2d
+            Vr_2d = self.radial_velocities_2d
+            Va_2d = self.axial_velocities_2d
+            
+            # local total velocity and resulting blade incident angle
+            U_2d   = np.sqrt(np.sqrt(Vt_2d**2 + Va_2d**2)**2 + Vr_2d**2) # (page 165 Leishman)
+            phi_2d = np.arctan(Va_2d/Vt_2d)     # (page 166 Leishman)
+            
+            # 2-D blade twist distribution 
+            theta_2d  = np.tile(total_blade_pitch,(Na ,1))
+            theta_2d  = np.repeat(theta_2d[np.newaxis,:, :], ctrl_pts, axis=0) 
+            
+            # 2-D chord distribution 
+            chord     = np.tile(c,(Na,1))  
+            chord_2d  = np.repeat(chord[np.newaxis,:, :], ctrl_pts, axis=0)    
+            
+            # local blade angle of attack
+            alpha  = theta_2d - phi_2d  # (page 166 Leishman)
+    
+            # 2-D atmospheric properties
+            a_2d  = np.tile(np.atleast_2d(a),(1,Nr))
+            a_2d  = np.repeat(a_2d[:, np.newaxis,  :], Na, axis=1)  
+            nu_2d       = np.tile(np.atleast_2d(nu),(1,Nr))
+            nu_2d       = np.repeat(nu_2d[:, np.newaxis,  :], Na, axis=1)    
+            
+            # local mach and reynolds numbers
+            Re         = (U_2d*chord_2d)/nu_2d  
+            Ma         = (U_2d)/a_2d    
+            
+            # aerodynamic calcs
+            Cl, Cdval = compute_aerodynamic_forces(a_loc, a_geo, cl_sur, cd_sur, ctrl_pts, Nr, Re, Ma, alpha, tc, Na)
+           
+            # More Cd scaling from Mach from AA241ab notes for turbulent skin friction 
+            T_2d    = np.tile(np.atleast_2d(T),(1,Nr))
```

Comment:
As much as possible you should share code with the one above

Reply:
I restructured, so the new version uses Drela's formulation completely but with the 2d analysis to allow for nonuniform inputs. This should be much cleaner now